### ISSUE TITLE
Inheritance with Self inferred types

### ISSUE THREAD

**Issue by mehdigmira** [microsoft/pyright#4670]
Hello,

I've just tried upgrading from 1.1.290 to the latest version, and encountered the following case.

```python
from __future__ import annotations


class A:
    def clone(self):  # inferred return type here is Self
        return self


class B(A):
    def clone(self):  # inferred return type here is B, should be Self
        return super().clone()


class C(B):
    def clone(self):  # inferred return type here is still B
        return super().clone()


reveal_type(C.clone)  # Type of "C.clone" is "(self: C) -> B"
reveal_type(C().clone())  # information: Type of "C().clone()" is "B"

```

I believe a better behaviour would be for C.clone return type to be inferred as "Self". 
Not sure whether this is a bug or and enhancement request, since inference is not smth described in a PEP.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.296, which I just published. It will also be included in a future release of pylance.
